app/src/main/java/ch/epfl/sdp/appart/FavoriteActivity.java

```diff
@@ -53,4 +77,71 @@ protected void onCreate(Bundle savedInstanceState) {
     private void updateList(List<Card> ls) {
         recyclerView.setAdapter(new CardAdapter(this, database, ls));
     }
+
+    private void saveLocally() {
+        localdb.cleanFavorites();
+        List<Card> cards = mViewModel.getFavorites().getValue();
+        for (Card c : cards) {
+            saveAd(c.getAdId(), c.getId());
+        }
+    }
+
+    private void saveAd(String adId, String cardId) {
+        User user = login.getCurrentUser();
+        database.getAd(adId)
+                .exceptionally(e -> {
+                    showFailToast();
+                    return null;
+                })
+                .thenAccept(ad -> {
+                    CompletableFuture<List<Bitmap>> images = getImages(adId, ad.getPhotosRefs());
+                    CompletableFuture<List<Bitmap>> panoramas = getImages(adId,
+                            ad.getPanoramaReferences());
+                    CompletableFuture<Bitmap> pfp = new CompletableFuture<>();
+                    database.accept(new GlideBitmapLoader(this, pfp,
+                            user.getProfileImagePathAndName()));
+                    images.thenAcceptBoth(panoramas, (imgs, panms) -> {
+                        pfp.exceptionally(e -> {
+                            showFailToast();
+                            return null;
+                        });
+                        pfp.thenAccept(pfpBitmap -> {
+                            localdb.writeCompleteAd(adId, cardId, ad, user, imgs, panms,
+                                    pfpBitmap);
+                        });
+                    }).exceptionally(e -> {

```

Comment:
Are the chain of future desired by you @rovati or you forgot to change it from yesterday ? Because at least for the first one we have a null pointer if there is an exception.

Reply:
I forgot a couple of them, fixed in the new PR I opened for auto login